### DISCUSSION TITLE
Add Custom WayPoints

### DISCUSSION THREAD

**Discussion by apppdevv2015** [owntracks/android#1760] [Q&A]
Hi Team,

Is there any way to add Custom waypoints or regions in Mobile app? I can see there is option to importWayPoints in mobile app... Please let me know what will be the format.  I tried xml and json for importing the waypoints.

**Comment by jpmens** [comment]
you tried xml and json? what data formats did you try? are you guessing?

waypoints can be added in the UI.

**Reply by apppdevv2015** [reply]
Yes, i have a requirements to add custom waypoints from the database.. Can you please let me know how it can be done?

Here is the example:

{
  "waypoints": [
    {
      "latitude": 37.7749,
      "longitude": -122.4194,
      "name": "San Francisco",
      "description": "City in California"
    },
    {
      "latitude": 34.0522,
      "longitude": -118.2437,
      "name": "Los Angeles",
      "description": "City in California"
    }
  ]
}

**Comment by jpmens** [comment]
This is [documented in the Booklet](https://owntracks.org/booklet/features/remoteconfig/#setwaypoints)

**Reply by apppdevv2015** [reply]
Thank you. Will it work for both IOS and Mobile app?  Is there any way to send data through MQTT and show on Mobile app?

**Comment by jpmens** [comment]
This method works for iOS and Android.

What data do you want to show on the apps?

**Reply by apppdevv2015** [reply]
I have requirements to create a regions in web application and show them in Mobile app through private server MQTT.. I sent custom waypoints (regions) to MQTT server and those are published but not added on mobile app under waypoints

**Reply by apppdevv2015** [reply]
Do we have import waypoints option in IOS? I didn't find it

**Comment by jpmens** [comment]
the message should be published with QoS>0

**Reply by apppdevv2015** [reply]
Okay so this is possible:

 so we can add custom waypoints (regions) to MQTT server through "owntracks/user/device" and will be displayed on mobile app under waypoints

**Reply by apppdevv2015** [reply]
I tried this:

 const region = {
   "_type": "cmd",
  "action": "setWaypoints",
  "waypoints": {
    "waypoints": [
      {
        "desc": "Aman2 place",
        "rad": 8867,
        "lon": 10.428771973,
        "lat": 46.935260881,
        "tst": 1437552714,
        "_type": "waypoint"
      }
    ],
    "_type": "waypoints"
  }
  }
    
    client.publish('owntracks/user/device', JSON.stringify(region), { qos: 1 }, (error) => {
        if (error) {
          console.error('Publish error:', error);
        } else {
          console.log('Region published');
        }
        client.end();
      });
      
      this is added on MQTT server but not added on Mobile app

**Reply by apppdevv2015** [reply]
I tried to import the below as json:

{
   "_type": "cmd",
  "action": "setWaypoints",
    "waypoints": {
        "waypoints": [
            {
                "desc": "Aman2 place",
                "rad": 8867,
                "lon": 10.428771973,
                "lat": 46.935260881,
                "tst": 1437552714,
                "_type": "waypoint"
            }
        ],
        "_type": "waypoints"
    }
  }
  
  But getting the message ImportFailed: Message is not a valid configuration
  
  Please help

**Comment by jpmens** [comment]
The topic is wrong. It should be `owntracks/user/device/cmd` and remote cmd must be enabled on the device.

**Reply by apppdevv2015** [reply]
Cmd option is not available on Andriod phone

**Reply by apppdevv2015** [reply]
Allowremote location and cmd in enabled in IOS device

**Reply by apppdevv2015** [reply]
Could you please send me the file sample to import through Importwaypoints? I tried the json file but it didn't work

**Reply by growse** [reply]
> Cmd option is not available on Andriod phone

Sure it is. What makes you think it's not?

**Reply by apppdevv2015** [reply]
Could you please send me the file sample to import through Importwaypoints in Andriod phone? I tried the json file but it didn't work

**Comment by jpmens** [comment]
This works for me.

```console
$ mosquitto_pub -u user -P password -t owntracks/user/device/cmd -f wp-blau.json -q 1
```

The file `wp-blau.json`:

```json
{
  "_type": "cmd",
  "action": "setWaypoints",
  "waypoints": {
    "_type": "waypoints",
    "waypoints": [
      {
        "_type": "waypoint",
        "tst": 1361636517,
        "rid": "blauw-utrecht-nl",
        "desc": "Restaurant Blauw",
        "rad": 50,
        "lat": 52.087822392366,
        "lon":  5.119438515394096
      }
    ]
  }
}
```

**Reply by apppdevv2015** [reply]
Is it working on Android? or IOS. Are you able to see it in Your Mobile app?

**Reply by apppdevv2015** [reply]
I tried it from Andriod phone, importwayPoints but  getting the message ImportFailed: Message is not a valid configuration.

Please let me know full process:

when i tried to send to MQTT Server and its published in server:

{
  "_type": "cmd",
  "action": "setWaypoints",
  "waypoints": {
    "_type": "waypoints",
    "waypoints": [
      {
        "_type": "waypoint",
        "tst": 1361636517,
        "rid": "blauw-utrecht-nl",
        "desc": "Restaurant Blauw",
        "rad": 50,
        "lat": 52.087822392366,
        "lon":  5.119438515394096
      }
    ]
  }
}

But not showing in Mobile app..

**Comment by jpmens** [comment]
Show me the COMMAND line you use to publish that to your MQTT broker. (Obviously without passwords.)

**Reply by apppdevv2015** [reply]
I just published this and its on mqtt server:

{ "_type": "cmd", "action": "setWaypoints", "waypoints": { "_type": "waypoints", "waypoints": [ { "_type": "waypoint", "tst": 1361636517, "rid": "blaur-utrecht-nl", "desc": "brr Blauw", "rad": 50, "lat": 52.087822392366, "lon": 5.119438515394096 } ] } }

But its not showing in Mobile app

**Reply by apppdevv2015** [reply]
I used code: 

const topic = 'owntracks/user/device/cmd'; // Replace with your topic
// Create MQTT client and connect
const client = mqtt.connect('mqtt server', {
    username: 'user',
    password: 'pass',
  });
  

// Read the waypoints file
const waypointsFile = 'wp-blau.json';
const waypoints = fs.readFileSync(waypointsFile, 'utf-8');

client.on('connect', () => {
    console.log('Connected to MQTT broker');
    
    // Publish the waypoints
    client.publish(topic, waypoints, { qos: 1 }, (err) => {
      if (err) {
        console.error('Failed to publish waypoints:', err);
      } else {
        console.log('Waypoints published successfully');
      }
      client.end();
    });
  });

**Comment by apppdevv2015** [comment]
![image](https://github.com/user-attachments/assets/b47ec3ac-9637-4dcd-b02b-e5685532663b)
This is published on MQTT Server but not showing in mobile

**Reply by apppdevv2015** [reply]
Hi @jpmens 

could you please let know what could be the issue in IOS device

**Comment by jpmens** [comment]
Don’t harass us please!

1. What is the setting for `cmd` on your iOS settings?
2. what is the content of `subTopic` on your iOS settings?
3. which MQTT broker are you using (brand of software)?
4. I asked you to show me the Linux command line with which you demonstrate that what I showed earlier works for you. You’ve not done that.
5. please stop showing me the content of the message that I give you; I know what’s in the Blauw waypoint description
6. show us logs of the MQTT broker at the point the `cmd` message arrives there
7. show us logs taken on the iOS app for the same time frame

**Comment by apppdevv2015** [comment]
Apologies Guys, Earlier I was tried the Andriod phone.. I will send you all the details soon

**Comment by apppdevv2015** [comment]
1) Attached the IOS Screen-shot
2) MQTT Broker : hivemq.cloud
3) MQTT LOG:
{ "_type": "cmd", "action": "setWaypoints", "waypoints": { "_type": "waypoints", "waypoints": [ { "_type": "waypoint", "tst": 13616356
![ImportedPhoto 743103272 602995](https://github.com/user-attachments/assets/f22a949e-157d-48ed-957b-add12c156330)
![ImportedPhoto 743103272 605077](https://github.com/user-attachments/assets/d5208f9b-41fc-4c2e-85f5-e4db8fbaaec6)
![ImportedPhoto 743103521 823017](https://github.com/user-attachments/assets/7ef6be7a-3fa4-4190-bdfe-057d7a67da93)
![ImportedPhoto 743103493 155285](https://github.com/user-attachments/assets/5f0ff143-d830-4cb1-bcc3-74fc6f9ea8b8)
17, "rid": "test-ui", "desc": "Ashwani Blauw", "rad": 50, "lat": 52.087822392366, "lon": 5.119438515394096 } ] } }

**Comment by jpmens** [comment]
that JSON that you constantly keep pasting is not a log …

The iOS screenshot shows the logfile name. You’ll need to send that to yourself or to us (support@owntracks.org) so we can see what’s IN the file

**Comment by apppdevv2015** [comment]
[org.mqttitude.MQTTitude 2024-07-19--11-10-56-739.log](https://github.com/user-attachments/files/16315511/org.mqttitude.MQTTitude.2024-07-19--11-10-56-739.log)

**Comment by jpmens** [comment]
in iOS settings, please make sure that in subTopic you see a topic which ends in `cmd` (I don’t see that from your screenshot)

**Reply by apppdevv2015** [reply]
do i just need to add " /cmd "

**Comment by apppdevv2015** [comment]
[org.mqttitude.MQTTitude 2024-07-19--11-10-56-739.log](https://github.com/user-attachments/files/16315511/org.mqttitude.MQTTitude.2024-07-19--11-10-56-739.log)

**Comment by jpmens** [comment]
also you’ve been publishing to a topic called `owntracks/user/test/cmd`, but in the screenshot it’s called `testUser` (capital U) and the device name is `C30…………`

The topic branch used by iOS MUST match the topic you’re publishing to.

**Comment by jpmens** [comment]
also you’ve been publishing to a topic called `owntracks/user/test/cmd`, but in the screenshot it’s called `User` (capital U) and the device name is `C30…………`

From your log I see you should be publishing to

```
owntracks/testUser/C30AE8FD-6135-4D6C-AD7B-B6333457ED2E/cmd
```

**Reply by apppdevv2015** [reply]
I did this now under SubTopic: owntracks/testUser/device/cmd

**Reply by apppdevv2015** [reply]
![ImportedPhoto 743105547 473611](https://github.com/user-attachments/assets/23a2872a-fd85-4582-ba9e-e2955edaaebe)

Still not showing under mobile regions

**Comment by jpmens** [comment]
That’s not what I asked you to do. Listen, this is getting very tedious for me. you are asking for help, and I’m trying to help you understand how to solve the issue. I said the topic to which you should MQTT publish; I did not say change the configuration of your phone.

I cannot keep track of what you’re doing.

I’m going to explain, and then I’ll stop: your program (javascript?) is trying to publish a payload to MQTT with a JSON which iOS will interpret. The phone is connected (hopefully) to a broker and your app to the same broker. If the broker permits, the publish your app does is read by the phone, and the phone will react. The topic to which your app publishes MUST MATCH the topic on which the phone subscribes.

This topic is configured automatically on the phone by setting up a user and a device. By default OwnTracks constructs topics of the form owntracks/user/device and for certain operations adds a branch to that, e.g. /cmd
Without this matching, the phone will never get what the app is publishing.

My reccomendation: Uninstall the iOS app and re-install it. Configure it with a username (e.g. jane) and a device name (e.g. myphone). The phone will then subscribe to owntracks/jane/myphone and owntracks/jane/myphone/cmd

When your app publishes (with QoS>0) the CMD payload to owntracks/jane/myphone/cmd the phone should show the newly received waypoint/region in the list of regions.

**Reply by apppdevv2015** [reply]
Okay jpmens. You are correct... Thank you so much for your help. The both is connected to same MQTT server and  attached the new setting.. testUser is username of MQTT so can't change now. I installed the app again and not changed the any subtopic here
![ImportedPhoto 743106708 055469](https://github.com/user-attachments/assets/c8344ec0-d2db-4fd0-9d6e-d7cdff79ddec)

My Code:

const topic = 'owntracks/testUser/myphone/cmd'; // Replace with your topic
// Create MQTT client and connect
const client = mqtt.connect('server', {
    username: 'testUser',
    password: 'pass',
  });
  

// Read the waypoints file
const waypointsFile = 'wp-blau.json';
const waypoints = fs.readFileSync(waypointsFile, 'utf-8');

client.on('connect', () => {
    console.log('Connected to MQTT broker');
    
    // Publish the waypoints
    client.publish(topic, waypoints, { qos: 1 }, (err) => {
      if (err) {
        console.error('Failed to publish waypoints:', err);
      } else {
        console.log('Waypoints published successfully');
      }
      client.end();
    });
  });

But still not showing in My Mobile... It could be frustrated for you but i really needs help here like what I am missing

**Reply by jpmens** [reply]
please show me a screenshot of the settings page with visible TrackerID, DeviceID, UserID 
and a screenshot of the subTopic and clientId lower down in EXPERT MODE please.

Also I need the OwnTracks log on iOS at the time you try to publish the cmd to it.

**Reply by apppdevv2015** [reply]
@jpmens  Its working, thank you very much for your help

However its working only for IOS and me, Topic: "owntracks/testUser/myphone/cmd ", Myphone is my deviceId but other users it could be different .Can I add the same regions to other users which are my under MQTT server. Also, please let me know about Android too and what settings are required.

**Reply by jpmens** [reply]
You might want to familiarize yourself with some of the OwnTracks concepts such as topics, etc. What you've now done for iOS will work likewise for Android, even though the buttons might be labeled differently.

**Reply by apppdevv2015** [reply]
for Andriod there is option : Remote Commands and I enabled it however its not showing on Android.. Also the regions are added only user device specific, Is there any way to add same regions for all users once?

**Reply by jpmens** [reply]
it works exactly the same way on Android; you have to adapt the topic to the user/device on each device.

**Reply by apppdevv2015** [reply]
@jpmens Thank you for your reply and I did the same however its not working on Android

**Reply by jpmens** [reply]
show me proof of it not working (settings, topics, etc)

**Reply by apppdevv2015** [reply]
@jpmens 

Thanks for your continuous support
I have attached the screen-shots. 

Please answer of below questions:
1) IOS Device: the regions are added only user device specific like (owntracks/jane/myphone/cmd), Is there any way to add same regions for all users device once? What would be the topic name?
2) How Import waypoints in Andriod device feature will work? I tried the sample you shared with me but its not working.
3) tst: Is it local time or UTC time? Please confirm
![1721797941274](https://github.com/user-attachments/assets/fdb55e02-921c-4656-be46-66df3b1c6b5a)

![1721797959641](https://github.com/user-attachments/assets/afc3f611-ec65-4114-9ac5-a11d89e7d9d6)
![1721797881820](https://github.com/user-attachments/assets/c9b15797-b49e-4cb5-b596-2cf684bf8fbb)

**Reply by jpmens** [reply]
1. no
2. you publish the data the same way you did for iOS, adjusting the topic
3. UTC

**Reply by apppdevv2015** [reply]
Okay, thanks 
1) In Android, i was talking to Import Waypoints feature. Will it work?
2) I have shared the configuration. Can you please check my configuration?

**Reply by jpmens** [reply]
1. sure it works, just give it a list of waypoints without the `cmd` bit in it

```json
{
  "_type": "waypoints",
  "waypoints": [
    {
      "_type": "waypoint",
      "tst": 1361636517,
      "rid": "blauw-utrecht-nl",
      "desc": "Restaurant Blauw",
      "rad": 50,
      "lat": 52.087822392366,
      "lon": 5.119438515394096
    }
  ]
}
```

2. no I cannot check your configuration because I don't know your environment, but what I see from the screenshots looks ok

**Reply by apppdevv2015** [reply]
Thanks Import file is working fine.. Is there any way for ios device to import the waypoints?

**Reply by jpmens** [reply]
how often are you going to ask that question? scroll up for the answers, please!!!
